File: ruby/lib/ter/runtime/environment.rb
# frozen_string_literal: true

module Ter
  module Runtime
    # Runtime environment for accessing validated variables
    class Environment
      attr_reader :schema, :metadata

      def initialize(schema, resolver)
        @schema = schema
        @resolver = resolver
        @resolved = resolver.resolve
        @metadata = @resolved[:metadata]

        unless @resolved[:valid]
          errors = @resolved[:errors].map { |e| "#{e[:field]}: #{e[:error]}" }
          raise "Environment validation failed: #{errors.join('; ')}"
        end

        @values = @resolved[:values]
      end

      # Get value by key (raises if not found)
      def get(key)
        raise "Variable '#{key}' not found" unless @values.key?(key)
        @values[key]
      end

      # Get value by key, returns nil if not found
      def get_optional(key)
        @values[key]
      end

      # Type-safe getters
      def get_string(key)
        value = get(key)
        raise "Expected string for #{key}, got #{value.class}" unless value.is_a?(String)
        value
      end

      def get_int(key)
        value = get(key)
        raise "Expected integer for #{key}, got #{value.class}" unless value.is_a?(Integer)
        value
      end

      def get_number(key)
        value = get(key)
        raise "Expected number for #{key}, got #{value.class}" unless value.is_a?(Float)
        value
      end

      def get_boolean(key)
        value = get(key)
        raise "Expected boolean for #{key}, got #{value.class}" unless [TrueClass, FalseClass].include?(value.class)
        value
      end

      def get_json(key)
        value = get(key)
        raise "Expected JSON (Hash/Array) for #{key}, got #{value.class}" unless value.is_a?(Hash) || value.is_a?(Array)
        value
      end

      # Check if variable is a secret
      def is_secret?(key)
        type = @schema.get_variable(key)
        type&.is_a?(Core::SecretType) || false
      end

      # Get metadata for variable
      def get_metadata(key)
        @metadata[key] || {}
      end

      # Get resolved values as hash
      def to_h
        result = {}
        @values.each do |key, value|
          result[key] = is_secret?(key) ? '[REDACTED]' : value
        end
        result
      end

      # Get all values (internal use)
      def to_h_internal
        @values.dup
      end

      # Validate environment
      def validate
        error_details = []
        @schema.get_variables.each do |name, type|
          value = @values[name]
          result = type.validate(value)
          error_details << "#{name}: #{result[:error]}" unless result[:valid]
        end

        { valid: error_details.empty?, errors: error_details }
      end
    end
  end
end

File: python/ter/runtime/environment.py
"""Type-safe environment access"""

from typing import Any, Optional, Dict
import json
from ter.core.schema import Schema
from ter.runtime.resolver import Resolver


class Environment:
    """Safe environment access with type validation"""
    
    def __init__(self, schema: Schema, resolver: Resolver):
        self.schema = schema
        self.resolver = resolver
        self._cache: Dict[str, Any] = {}
    
    def get(self, name: str, default: Any = None) -> Any:
        """Get raw value"""
        resolved = self.resolver.resolve(name)
        if resolved:
            return resolved.value
        return default
    
    def get_string(self, name: str) -> str:
        """Get as string"""
        value = self.get(name)
        if value is None:
            raise ValueError(f"{name} is required")
        return str(value)
    
    def get_int(self, name: str) -> int:
        """Get as integer"""
        value = self.get(name)
        if value is None:
            raise ValueError(f"{name} is required")
        return int(value)
    
    def get_float(self, name: str) -> float:
        """Get as float"""
        value = self.get(name)
        if value is None:
            raise ValueError(f"{name} is required")
        return float(value)
    
    def get_bool(self, name: str) -> bool:
        """Get as boolean"""
        value = self.get(name)
        if value is None:
            raise ValueError(f"{name} is required")
        
        if isinstance(value, bool):
            return value
        
        if isinstance(value, str):
            return value.lower() in ("true", "1", "yes", "on")
        
        return bool(value)
    
    def get_json(self, name: str) -> Any:
        """Get as JSON"""
        value = self.get(name)
        if value is None:
            raise ValueError(f"{name} is required")
        
        if isinstance(value, (dict, list)):
            return value
        
        return json.loads(str(value))
    
    def get_optional(self, name: str) -> Optional[Any]:
        """Get optional value"""
        return self.get(name)
    
    def validate(self) -> Dict[str, Any]:
        """Validate all variables"""
        values = {}
        for name in self.schema.get_variables():
            resolved = self.resolver.resolve(name)
            if resolved:
                values[name] = resolved.value
        
        return self.schema.validate(values)
    
    def get_all_values(self) -> Dict[str, Any]:
        """Get all resolved values"""
        all_values = {}
        for name, var_def in self.schema.get_variables().items():
            value = self.get(name)
            if value is not None:
                all_values[name] = value
        
        return all_values
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        return self.get_all_values()
    
    def get_metadata(self, name: str) -> Optional[Dict[str, Any]]:
        """Get metadata about a value"""
        resolved = self.resolver.resolve(name)
        if resolved:
            return {
                "source": resolved.source,
                "resolved_at": resolved.resolved_at.isoformat()
            }
        return None
